doc/mkhexgrid.1
.\" Process this file with
.\" groff -man -Tascii mkhexgrid.1
.\"
.TH MKHEXGRID 1 "January 2007" "mkhexgrid 0.1.1"

.SH NAME
mkhexgrid \- make hexagonal grids

.SH SYNOPSIS
.B mkhexgrid
.RI [ OPTION "]... [" specfile "[ " outfile ]]

.SH DESCRIPTION
mkhexgrid creates hexagonal grids in PNG, PostScript, and Scalable Vector
Graphics (SVG) formats, optionally labeled with coordinates and hex centers.
.I specfile
is the name of the file containing hex grid parameters and
.I outfile
is the name of the file into which output is written. If
.I outfile
is not specified, output will be written to standard output. If
.I specfile
is not specified, grid parameters may be read from standard input.

.SH OPTIONS
.SS Option Syntax

mkhexgrid uses GNU getopt to process command-line arguments. Each option may be given as
.B --option-name
for options which take no arguments, or as
\fB--option-name\fR=\fIargument\fR
for options which take arguments.

Lengths and sizes are in pixels for PNG and SVG output and are given without specifying the unit. Lengths and sizes may be specified in inches, millimeters, or points (as in, mm, or pt) for PostScript output, and default to points when no unit is given. (1in = 72pt = 25.4mm) Lengths and sizes for PNG output must be integers, with the exception of coordinate size, which may be a decimal. All angles are given as decimals in degrees.

Colors are given in hexadecimal RRGGBB format for PNG and SVG output (e.g., FFFFFF is white, 8C6AC6 is a shade of purple). Colors are given as R,G,B triples for PostScript output, where R, G, and B are values in the interval [0,1] (e.g., 1,0,0 is red, and 0.8,0.8,0.8 is a light grey).

Opacity ranges over integers in [0,127] for PNG output, reals in [0,1] for SVG output, and is ignored for PostScript output.

.SS Output Options
.TP
.B --antialias
Turn on antialiased output. This option has effect for PNG output only, since PostScript and SVG are antialiased by nature.

.TP
\fB-o\fR \fIfilename\fR, \fB--outfile\fR=\fIfilename\fR
Write output to the file called \fIfilename\fR. The default is to print to standard output if no output filename is given or if a dash (-) is given as the filename. To write to a file named -, give ./- as the filename.

.TP
\fB--output\fR=\fItype\fR
Set the output type to \fItype\fR. Permissible values are 'png' for PNGs, 'ps' for PostScript, and 'svg' for SVG.

.SS Grid Options
Not all of hex width, hex height, hex side, image width, image height, rows, and columns need be given in order to draw a grid. A grid will be drawn so long as enough of these are specified so that the ones omitted may be calculated.

.TP
\fB--hex-width\fR=\fIlength\fR
Set the width of hexes to \fIlength\fR.

.TP
\fB--hex-height\fR=\fIlength\fR
Set the height of hexes to \fIlength\fR.

.TP
\fB--hex-side\fR=\fIlength\fR
Set the length of hex sides to \fIlength\fR.

.TP
\fB--image-width\fR=\fIlength\fR
Set the width of the image to \fIlength\fR.

.TP
\fB--image-height\fR=\fIlength\fR
Set the height of the image to \fIlength\fR.

.TP
\fB--image-margin\fR=\fIm\fR, \fB--image-margin\fR=\fIt,r,b,l\fR
The first form sets all four margins to \fIm\fR. The second form proceeds clockwise to set the top, right, bottom, and left margins individually to \fIt\fR, \fIr\fR, \fIb\fR, and \fIl\fR, respectively.

.TP
\fB--centered\fR
Center grid with respect to the image boundaries less the image margins. Useful when rows and columns are calculated rather than specified.

.TP
\fB--rows\fR=\fIn\fR
Draw \fIn\fR rows of hexes. \fIn\fR must be a positive integer.

.TP
\fB--columns\fR=\fIn\fR
Draw \fIn\fR columns of hexes. \fIn\fR must be a positive integer.

.TP
\fB--grid-color\fR=\fIcolor\fR
Set the color of the grid lines. Default is 50% grey: 808080 for PNG and SVG output, 0.5,0.5,0.5 for PostScript output.

.TP
\fB--grid-opacity\fR=\fIopacity\fR
Set the opacity of the grid lines. Defaults to fully opaque: 0 for PNG output, 1 for SVG output, and is ignored for PostScript output.

.TP
\fB--grid-thickness\fR=\fIsize\fR
Set the thickness of the grid lines to \fIsize\fR. Defaults to 1px for PNG and SVG output, and 1pt for PostScript. This option is ignored for antialiased PNG output, due to limitations of the underlying GD drawing library.

.TP
\fB--grid-grain\fR=\fIgrain\fR
Set the grain of the hex grid. Permissible values are 'h' for horizontal and 'v' for vertical. With horizontal grain, the rows are straight and the columns wavy; vertical grain is the opposite, and is the default.

.TP
\fB--grid-start\fR=\fIstart\fR
If the grain is vertical, set whether the first column starts in or out. If the grain is horizontal, set whether the first row starts in or out. Permissible values are 'i' for in, or 'o' for out. Which column or row is the first is fixed by the coordinate origin.

.SS Coordinate Options
These options affect the appearance and placement of coordinates within hexes.

.TP
\fB--coord-color\fR=\fIcolor\fR
Set the color of the hex coordinates. Default is 50% grey: 808080 for PNG and SVG output, 0.5,0.5,0.5 for PostScript output.


.TP
\fB--coord-opacity\fR=\fIopacity\fR
Set the opacity of the hex coordinates. Defaults to fully opaque: 0 for PNG output, 1 for SVG output, and is ignored for PostScript output.

.TP
\fB--coord-format\fR=\fIformat\fR
Set the format of the hex coordinates. The format string syntax is similar to that used by the
.BR printf (3)
function in C. The following special markers cause values to be substituted into the format string where they appear:

    \fB%c\fR
        Prints the hex column as a number.
    \fB%r\fR
        Prints the hex row as a number.
    \fB%C\fR
        Prints the hex column as a capital letter.
    \fB%R\fR
        Prints the hex row as a capital letter.

The field width can be specified for numeric types by inserting a number between the % and the format letter; numeric types can be zero-padded by prepending a 0 to the field width. The alphabetic formats count AB as the successor of AA by default; tally-like counting where BB is the successor of AA can be specified by inserting a t before the format letter. A percent sign may be specified by %%. The default format is %02c%02r. Coordinates may be disabled by giving an empty string ("") as the format.

.TP
\fB--coord-font\fR=\fIfont\fR
Set the font for coordinate text.

.TP
\fB--coord-size\fR=\fIsize\fR
Set the size, in points, of the coordinate text. Defaults to 8px for SVG, 8pt for PNG and PostScript. This is the only length or size which is measured in points for PNG output.

.TP
\fB--coord-bearing\fR=\fItheta\fR
Place coordinates at \fItheta\fR degrees counterclockwise from the hex center. The default bearing is 90 degrees, vertically above the hex center.

.TP
\fB--coord-distance\fR=\fIlength\fR
Place coordinates \fIlength\fR away from the hex center. Defaults to 0.

.TP
\fB--coord-tilt\fR=\fItheta\fR
Tilt coordinates at \fItheta\fR degrees counterclockwise from the horizontal. The default tilt is 0 degrees, which yields horizontal text.

.TP
\fB--coord-column-skip\fR=\fIn\fR
Print coordinates only for columns divisible by \fIn\fR. The default is 1, which prints coordinates for all columns.

.TP
\fB--coord-row-skip\fR=\fIn\fR
Print coordinates only for rows divisible by \fIn\fR. The default is 1, which prints coordinates for all rows.

.TP
\fB--coord-column-start\fR=\fIn\fR
Number columns starting from \fIn\fR. The default is 1.

.TP
\fB--coord-row-start\fR=\fIn\fR
Number rows starting from \fIn\fR. The default is 1.

.TP
\fB--coord-origin\fR=\fIcorner\fR
Set location of coordinate origin. Permissible locations are 'ul' for upper left, 'ur' for upper right, 'll' for lower left, and 'lr' for lower right. Upper left is the default.

.SS Hex Center Options

.TP

.TP
\fBcenter-style\fR=\fIstyle\fR
Set style of hex centers to \fIstyle\fR. Permissible styles are 'n' for none, 'd' for dots, and 'c' for crosses.

.TP
\fBcenter-color\fR=\fIcolor\fR
Set the color of hex centers. Default is 50% grey: 808080 for PNG and SVG output, 0.5,0.5,0.5 for PostScript output.

.TP
\fBcenter-opacity\fR=\fIopacity\fR
Set the opacity of hex centers. Defaults to fully opaque: 0 for PNG output, 1 for SVG output, and is ignored for PostScript output.

.TP
\fBcenter-size\fR=\fIsize\fR
Set the size of hex centers to \fIsize\fR. Defaults to 3px for PNG and SVG, 3pt for PostScript. Center size must be an integer for PNG output.

.SS Background Options

.TP
\fB--bg-color\fR=\fIcolor\fR
Set the background color. Defaults to FFFFFF for PNG output, and to none for SVG and PostScript output.

.TP
\fB--bg-opacity\fR=\fIopacity\fR
Set the background opacity. Default is fully opaque: 0 for PNG output, 1 for SVG output, and is ignored for PostScript output.

.TP
\fB--matte\fR
Prevent the background from extending into the image margins.

.SS Miscellaneous Options

.TP
\fB--\fR
Terminates the list of options. Arguments which appear after a \fB--\fR will not be interpreted as options. This is useful if you want to read a specfile the name of which begins with two dashes.

.TP
\fB--help\fR
Display help and exit.

.TP
\fB--version\fR
Display version information and exit.

.SS Specfile Syntax

Due to the large number of options which may be necessary to generate a single hex grid, users may also specify hex grid parameters in a specfile which can be read from standard input or named as the first non-option argument on the command line. A specfile contains =-separated option/value pairs, one pair per line. Comments may be preceded by a #, either following an option/value pair or on lines by themselves. Values containing spaces must be double-quoted. Double quotation marks may be included in quoted values by preceding them with a backslash. Options which require no value on the command line must be given a dummy value in a specfile. Options given on the command line override options given in a specfile.

Example:

   # specfile example
   output = png
   hex-side = 30     # this is an end-of-line comment
   coord-format = "%C\\" %02r"
   antialias = true  # 'true' is a dummy value 

Here, the coordinate format will produce coordinates like A" 01, which isn't something anyone is likely to want, but it illustrates how to include spaces and quotation marks in a quoted value.

.SH EXAMPLES

Fill a 22"x34" sheet with 0.5" hexes, horizontal grain:

.RS 4
mkhexgrid --output=ps --image-width=34in --image-height=22in --hex-side=0.5in --coord-bearing=0 --coord-dist=0.3in --coord-size=8 --grid-thickness=1 --coord-font=Helvetica --grid-grain=h --grid-start=o --coord-tilt=-90 --centered -o test.ps

.SH AVAILABILITY

mkhexgrid is available from 

         http://www.nomic.net/~uckelman/mkhexgrid.

New releases of mkhexgrid are announced on the
      
         mkhexgrid-announce@nomic.net

mailing list. To join, send a message to

         mkhexgrid-announce-request@nomic.net

with the word \fIsubscribe\fR in the message body.

.SH AUTHOR

Written by Joel Uckelman <uckelman@nomic.net>.
 
.SH BUGS

Please report bugs to <mkhexgrid-bugs@nomic.net>.

.SH COPYRIGHT

Copyright \(co 2006, 2007 Joel Uckelman.

This is free software. You may redistribute copies of it under the terms of the GNU General Public License. There is NO WARRANTY, to the extent permitted by law.


HISTORY
mkhexgrid 0.1.1, 10 January 2007

* Now builds properly on Mac OS X.
* "%c" and "%r" coordinate formats are now rendered properly.
* Coordinate bearing now defaults to 90 degrees instead of 0. 
* Various corrections in the documentation.
* Thanks to Steve Huff and Masaki Nakazawa for finding the bugs corrected
  in this release.

mkhexgrid 0.1.0, 6 August 2006
* Initial release.


mkhexgrid.cpp
// $Id$
/* mkhexgrid -- generates hex grids
 * Copyright (C) 2006 Joel Uckelman
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

#include <cstdlib>
#include <exception>
#include <iostream>
#include <fstream>
#include <map>
#include <stdexcept>
#include <string>
using namespace std;

#include <boost/lexical_cast.hpp>
using namespace boost;

#include <getopt.h>

#include "grid.h"

void parse_spec(istream &in, map<string, string> &opt);
void print_help();

struct option long_options[] = {
   { "hex-width",          1, 0, 0 },
   { "hex-height",         1, 0, 0 },
   { "hex-side",           1, 0, 0 },
   { "image-width",        1, 0, 0 },
   { "image-height",       1, 0, 0 },
   { "image-margin",       1, 0, 0 },
   { "rows",               1, 0, 0 },
   { "columns",            1, 0, 0 },
   { "grid-color",         1, 0, 0 },
   { "grid-opacity",       1, 0, 0 },
   { "grid-thickness",     1, 0, 0 },
   { "grid-grain",         1, 0, 0 },
   { "grid-start",         1, 0, 0 },
   { "coord-color",        1, 0, 0 },
   { "coord-opacity",      1, 0, 0 },
   { "coord-format",       1, 0, 0 },
   { "coord-font",         1, 0, 0 },
   { "coord-size",         1, 0, 0 },
   { "coord-bearing",      1, 0, 0 },
   { "coord-distance",     1, 0, 0 },
   { "coord-tilt",         1, 0, 0 },
   { "coord-column-skip",  1, 0, 0 },
   { "coord-row-skip",     1, 0, 0 },
   { "coord-column-start", 1, 0, 0 },
   { "coord-row-start",    1, 0, 0 },
   { "coord-origin",       1, 0, 0 },
   { "bg-color",           1, 0, 0 },
   { "bg-opacity",         1, 0, 0 },
   { "matte",              0, 0, 0 },
   { "center-style",       1, 0, 0 },
   { "center-color",       1, 0, 0 },
   { "center-opacity",     1, 0, 0 },
   { "center-size",        1, 0, 0 },
   { "antialias",          0, 0, 0 },
   { "centered",           0, 0, 0 },
   { "output",             1, 0, 0 },
   { "outfile",            1, 0, 'o' },
   { "help",               0, 0, 'h' },
   { "version",            0, 0, 'v' },
   { 0, 0, 0, 0 }
};


int main(int argc, char **argv)
{
   map<string, string> opt;

   int c;
   int option_index = 0;

   while (1) {
      c = getopt_long(argc, argv, "i:o:", long_options, &option_index);
      if (c == -1) break;  // end of options

      switch (c) {
      case 0:
         opt[long_options[option_index].name] = optarg ? optarg : "";
         break;
      case 'i':
         opt["infile"] = optarg;
         break;
      case 'o':
         opt["outfile"] = optarg;
         break;
      case 'v':
         cout << "mkhexgrid version " << VERSION << endl;
         exit(0);
      case 'h':
         print_help();
         exit(0);
      default:
         exit(1);    // getopt produces an error message
      }
   }

   try {
      // get name of specfile, if given as a positional parameter
      if (optind < argc) {
         if (opt.find("infile") != opt.end())
            throw runtime_error("specfile specified twice on command line");
         
         opt["infile"] = argv[optind++];

         // get name of outfile, if given as a positional parameter
         if (optind < argc) {
            if (opt.find("outfile") != opt.end()) 
             throw runtime_error("output file specified twice on command line");
            opt["outfile"] = argv[optind++];

            if (optind < argc) throw runtime_error("too many arguments");
         }
      }

      // read and parse specfile
      map<string,string>::const_iterator i = opt.find("infile");
      if (i != opt.end()) {
         if (i->second == "-") parse_spec(cin, opt); 
         else {
            ifstream in;
            in.open(i->second.c_str(), ios::in | ios::binary);
            if (!in)
             throw runtime_error("cannot read spec file `" + i->second + "'");
            parse_spec(in, opt);
            in.close();
         }
      } 
   
      // draw hex grid
      Grid g(opt);
      g.draw();
   }
   catch (std::exception &e) {
      cerr << argv[0] << ": " << e.what() << endl;
      exit(1);
   }

   return 0;
}


void parse_spec(istream &in, map<string, string> &opt)
{
   string key, val;
   unsigned int line = 1, col = 0;
   char c;
   enum { START, KEY, PRE_EQ, POST_EQ, VAL, QVAL, GUARD, END }
      state = START;
   
   while (in.get(c)) {
      ++col;
      switch (state) {
      case START:
         switch (c) {
         case '\n':
            ++line;
            col = 0;
         case ' ':
         case '\t':
         case '\r':
            break;
         case '#':
            do { 
               in.get(c);
               if (c == '\n') {
                  ++line;
                  col = 0;
                  break;
               }
            } while (in);
            break;
         default:
            key += c;
            state = KEY;
            break;
         }
         break;
      case KEY:
         switch (c) {
         case ' ':
         case '\t':
         {
            unsigned int i = 0;
            for ( ; long_options[i].name &&
                    long_options[i].name != key; ++i);
            if (!long_options[i].name)
               throw runtime_error("unrecognized option `" + key +
                  "' in specfile");
            state = PRE_EQ;
            break;
         }
         case '\r':
         case '\n':
         case '#':
            throw runtime_error("malformed spec file at"
               " line " + lexical_cast<string>(line) + ","
               " column " + lexical_cast<string>(col));
            break;
         case '=':
         {
            unsigned int i = 0;
            for ( ; long_options[i].name &&
                    long_options[i].name != key; ++i);
            if (!long_options[i].name)
               throw runtime_error("unrecognized option `" + key +
                  "' in specfile");
            state = POST_EQ;
            break;
         }
         default:
            key += c;
            break;
         }
         break;
      case PRE_EQ:
         switch (c) {
         case ' ':
         case '\t':
            break;
         case '=':
            state = POST_EQ;
            break;
         default:
            throw runtime_error("malformed spec file at"
               " line " + lexical_cast<string>(line) + ","
               " column " + lexical_cast<string>(col));
         }
         break;
      case POST_EQ:
         switch (c) {
         case ' ':
         case '\t':
            break;
         case '\r':
         case '\n':
         case '#':
            throw runtime_error("malformed spec file at"
               " line " + lexical_cast<string>(line) + ","
               " column " + lexical_cast<string>(col));
         case '"':
            state = QVAL;
            break;
         default:
            val += c;
            state = VAL;
            break;
         }
         break;
      case QVAL:
         switch (c) {
         case '"':
            if (opt.find(key) == opt.end()) opt[key] = val;
            key.clear();
            val.clear();
            state = END;
            break;
         case '\\':
            state = GUARD;
            break;
         case '\n':
            throw runtime_error("malformed spec file at"
               " line " + lexical_cast<string>(line) + ","
               " column " + lexical_cast<string>(col));
         default:
            val += c;
            break;
         }
         break;
      case GUARD:
         switch (c) {
         case '\\':
            val += "\\\\";
            break;
         case '"':
            val += '"';
            break;
         case '\n':
            throw runtime_error("malformed spec file at"
               " line " + lexical_cast<string>(line) + ","
               " column " + lexical_cast<string>(col));
         default:
            val += "\\" + c;
            break;
         }
         state = QVAL;
         break;
      case VAL:
         switch (c) {
         case '=':
         case '#':
            throw runtime_error("malformed spec file at"
               " line " + lexical_cast<string>(line) + ","
               " column " + lexical_cast<string>(col));
         case '\n':
            ++line;
            col = 0;
            if (opt.find(key) == opt.end()) opt[key] = val;
            key.clear();
            val.clear();
            state = START;
            break;
         case '\r':
         case '\t':
         case ' ':
            if (opt.find(key) == opt.end()) opt[key] = val;
            key.clear();
            val.clear();
            state = END;
            break;
         default:
            val += c;
            break;
         }
         break;
      case END:
         switch (c) {
         case '\n':
            ++line;
            col = 0;
            state = START;
            break;
         case ' ':
         case '\t':
         case '\r':
            break;
         case '#':
            do { 
               in.get(c);
               if (c == '\n') {
                  ++line;
                  col = 0;
                  break;
               }
            } while (in);
            state = START;
            break;
         default:
            throw runtime_error("malformed spec file at"
               " line " + lexical_cast<string>(line) + ","
               " column " + lexical_cast<string>(col));
         }
         break;
      }
   }
   
   switch (state) {
   case START:
   case END:
      break;
   case QVAL:
   case GUARD:
   case KEY:
   case PRE_EQ:
   case POST_EQ:
      throw runtime_error("malformed spec file at"
         " line " + lexical_cast<string>(line) + ","
         " column " + lexical_cast<string>(col));
   case VAL:
      if (opt.find(key) == opt.end()) opt[key] = val;
      break; 
   }
}


void print_help()
{
   cout <<
"Usage: mkhexgrid [OPTION]... [SPECFILE [OUTFILE]]\n"
"Make hexagonal grids.\n"
"\n"
"Options:\n"
"   --hex-side=LENGTH        set hex sdie to LENGTH\n"
"   --hex-width=LENGTH       set hex width to LENGTH\n"
"   --hex-height=LENGTH      set hex height to LENGTH\n"
"   --columns=N              set number of hex columns to N\n"
"   --rows=N                 set number of hex rows to N\n"
"   --image-width=LENGTH     set image width to LENGTH\n"
"   --image-height=LENGTH    set image height to LENGTH\n"
"   --margin=LENGTH          set image margins to LENGTH\n"
"   --margin=T,R,B,L         set image margins independently\n"
"   --grid-color=COLOR       set grid color to COLOR\n"
"   --grid-opacity=OPACITY   set grid opacity to OPACITY\n"
"   --grid-width=SIZE        set width of grid lines to SIZE\n" 
"   --grid-grain=G           set grid grain, G = h, v\n"    
"   --grid-start=S           set first column/row to start in (i) or out (o)\n"
"   --coord-color=COLOR      set coordinates color to COLOR\n"
"   --coord-opacity=OPACITY  set coordinates opacity to OPACITY\n"
"   --coord-format=FORMAT    set coordinates format to FORMAT\n"
"   --coord-font=FONT        set coordinates font to FONT\n"
"   --corod-size=SIZE        set coordinates font size to SIZE\n"
"   --coord-bearing=ANGLE    place coordinates at ANGLE from hex centers\n"
"   --coord-distance=LENGTH  place coordinates LENGTH from hex centers\n"
"   --coord-tilt=ANGLE       tilt coordinates by ANGLE\n"
"   --coord-column-skip=N    number columns divisible by N\n"
"   --coord-row-skip=N       number rows divisible by N\n"
"   --coord-column-start=N   number columns starting from N\n"
"   --coord-row-start=N      number rows starting from N\n"
"   --coord-origin=O         set coordinate origin O = ul, ur, ll, lr\n"
"   --bg-color=COLOR         set background color to COLOR\n"
"   --bg-opacity=OPACITY     set background opacity to OPACITY\n"
"   --matte                  clip background at margins\n"               
"   --center-style=STYLE     set hex center style STYLE = n, d, c\n"
"   --center-color=COLOR     set hex center color to COLOR\n"
"   --center-size=SIZE       set hex center size to SIZE\n"
"   --antialias              turn on antialiased output\n"
"   --centered               center grid within the image margins\n"
"   --output=TYPE            set output TYPE = png, ps, svg\n"
"-o --outfile=FILE           set output filename to FILE\n"
"   --help                   display this help and exit\n"
"   --version                display version information and exit\n"
"\n"
"Please visit http:://www.nomic.net/~uckelman/mkhexgrid for updates and bug\n"
"reports.\n";
}
